26 create a script to make predictions at country level

soriadelva opened the pull request:
This PR contains all adjustments that will make the wiSDM workflow operational again. Except from fixing a couple of smaller errors, the functionality of the wiSDM workflow was kept largely the same.
The README file has been updated, providing all necessary information to run this workflow.

Shortly, run the following scripts (stored in the `src` folder) in the designated order to test the wiSDM workflow

0. **Script 00_prepare_files_and_folders**: Sets up the folder structure and downloads the files (climate rasters, habitat predictors, spatial boundaries,...) necessary to run the workflow.
1. **Script 01_global_occurrence_download.R**: After specifying the name of your project and your species of interest, this script creates a project folder on your local computer and retrieves occurrence data for the respective species from the Global Biodiversity Information Facility (GBIF). To allow this data download, a pop-up will appear, requesting you to enter your GBIF username, password, and email address.
2. **Script 02_fit_global_model.R**: Builds a global-scale climate-only species distribution model (SDM) for each species specified in script 01. Ensure you use the same project name as in the former script.
3. **Script 03_fit_European_model.R**: Generates European-level SDMs for the specified species. Again, use the same project name as in script 01.
4. **Script 04_Make_country_level_predictions.R**: Predicts species distributions and generates confidence maps under different climate change scenarios (RCP 2.6, RCP 4.5, and RCP 8.5) for a country or region of interest. At the moment, the workflow is only operational for Belgium, but this will be adjusted soon to incorporate more countries.
<br>

SanderDevisscher commented on src/00_prepare_files_and_folders.R:
```diff
@@ -0,0 +1,429 @@
+#--------------------------------------------
+#-----------  Load packages  ----------------
+#--------------------------------------------
+packages <- c("curl", "zen4R", "here", "sf", "terra")
```
Maybe add `rnaturalearth` to the list of packages. At least for me it interupted the flow.

SanderDevisscher commented on src/01_global_occurrence_download.R:
```diff
@@ -0,0 +1,194 @@
+
+#Time period
+year_begin <- 1971
+year_end <-2010
```
Why 2010 ? Is this specific for frederiques case ? has this to do with climate layers ? ...?

soriadelva replied:
This was the default in wiSDM. The paper says the following: The default
end dates of 1971 and 2010 were chosen to maximize the number
of available observations while staying with +/- 5 years of the end
dates used for the climate data to minimize a temporal mismatch
between the two datasets (

SanderDevisscher replied:
Do you still use the same climate layers ? ea does the most recent climate layer still end at approx 2015 ?

soriadelva replied:
Yes, atm we still use the original CHELSA layers used during the creation of the wiSDM workflow and the RMI corrected layers created by the KMI for TrIAS. However, together with researchers from OneSTOP and GuardIAS we agreed that one of the future enhancements is to run the workflow on the future climate layers. Until then, we still use the old ones though.

SanderDevisscher replied:
Ok in that case leave as is untill new layers are used

SanderDevisscher commented on src/02_fit_global_model.R:
I get the following error:
```
Folder 'Rasters/Interim' created at path: './data/projects/Test_Xenopus_Lithobates/Lithobates_catesbeianus_2427091/Rasters/Interim' 🎉
Folder 'Rasters/Global' created at path: './data/projects/Test_Xenopus_Lithobates/Lithobates_catesbeianus_2427091/Rasters/Global' 🎉
Folder 'PDFs' created at path: './data/projects/Test_Xenopus_Lithobates/Lithobates_catesbeianus_2427091/PDFs' 🎉
Folder 'PNGs' created at path: './data/projects/Test_Xenopus_Lithobates/Lithobates_catesbeianus_2427091/PNGs' 🎉
  |==============================================================                                                             |  50%
Progress interrupted by simpleError condition: [mask] extents do not match
Error: [mask] extents do not match
```

SanderDevisscher replied:
I'll try and pinpoint where it goes wrong

SanderDevisscher commented on src/02_fit_global_model.R:
```diff
@@ -0,0 +1,568 @@
+} else {
+  devtools::install_github("zachmayer/caretEnsemble@2.0.3")
+  library(caretEnsemble)
+  rm(current_version, desired_version)
```
`current_version` is not determined in this else statement since `caretEnsemble` is not yet installed

soriadelva replied:
> `current_version` is not determined in this else statement since `caretEnsemble` is not yet installed

Fixed 👍

SanderDevisscher commented on src/02_fit_global_model.R:
```diff
@@ -0,0 +1,568 @@
+#--------------------------------------------
+#-------Start loop for SDM modelling --------
+#--------------------------------------------
+with_progress({
```
The error occurs in this part of the script

SanderDevisscher commented on src/02_fit_global_model.R:
```diff
@@ -0,0 +1,568 @@
+    
+    #Mask biasgrid with one of the climatic layers, to make sure it doesn't extend beyond them
+    climategrid_rast<-terra::crop(globalclimpreds_terra[[1]], wwf_ecoSub1_ext)
+    biasgrid_sub<-terra::mask(biasgrid_sub, climategrid_rast) 
```
The error occurs in this step

SanderDevisscher replied:
Testing whether updating `terra` from `1.8-29` to `1.8-42` fixes the issue

SanderDevisscher replied:
error remains 😢

soriadelva replied:
I'll have a look!

SanderDevisscher replied:
I suspect the extent and resolution of the rasters is a bit unaligned 
```
> terra::ext(biasgrid_sub)
SpatExtent : -160, 142, -44, 82 (xmin, xmax, ymin, ymax)
> terra::ext(climategrid_rast)
SpatExtent : -159.71680563665, 142.06652648955, -44.22513907195, 82.34986042175 (xmin, xmax, ymin, ymax)
> terra::res(biasgrid_sub)
[1] 1 1
> terra::res(climategrid_rast)
[1] 0.008333333 0.008333333
```

Maybe you can use `terra::resample()` to fix it!

soriadelva replied:
Thanks! I checked it and the problem seems to be specific to the biasgrids of reptiles, amphibians, molluscs and mammals. Those of plants and birds are okay. I'll add logic that resamples the former biasgrids.

SanderDevisscher replied:
offcourse I'm testing with _Lithobates catesbeianus_  😅

soriadelva replied:
Fixed it 👍 Thanks Sander!

SanderDevisscher replied:
ok to test ?

soriadelva replied:
Go for it! 👍

SanderDevisscher replied:
The issue seems to be fixed! 🎉

SanderDevisscher commented on src/02_fit_global_model.R:
```diff
@@ -0,0 +1,574 @@
+#--------------------------------------------
+options("rgdal_show_exportToProj4_warnings"="none")
+
+packages <- c( "dplyr", "stringr", "here", "qs","CoordinateCleaner","terra", "raster", "rnaturalearth", "rnaturalearthdata", 
```
add `gbm` to packages

soriadelva replied:
Done 👍

SanderDevisscher replied:
also add  `PresenceAbscence`

soriadelva replied:
Done 👍

SanderDevisscher commented on src/04_Make_country_level_predictions.R:
```diff
@@ -0,0 +1,825 @@
+#---------- Define resampling raster -------------
+#-------------------------------------------------
+#Used as a template when global model predictions need to be in same raster format as those for the EU model
+resampling_raster<-rast(here("./data/external/climate/byEEA_finalRCP/belgium_rcps/rcp26/anngdd100.tif"))
```
This file seems to be missing from the setup. 

```
Error: [rast] file does not exist: C:/Users/sander_devisscher/Documents/GitHub/risk-modelling-and-mapping/./data/external/climate/byEEA_finalRCP/belgium_rcps/rcp26/anngdd100.tif
In addition: Warning message:
C:/Users/sander_devisscher/Documents/GitHub/risk-modelling-and-mapping/./data/external/climate/byEEA_finalRCP/belgium_rcps/rcp26/anngdd100.tif: No such file or directory (GDAL error 4) 
```

SanderDevisscher replied:
2 issues "./" does some weird shit with `here()` and `anngdd100.tif == var10_30yrMeanAnnualCumulatedGDDAbove5degreesC_rcp26_2041_2070_Belgium.tif`. At least that how I understand it.

damianooldoni replied:
`here::here("data", "external", "climate", "byEEA_finalRCP", "rcp26", "anngdd100.tif")`

amyjsdavis commented:
you should be able to match the extent of the dist2water raster to the
other habitat rasters in R so you don't have to exclude it.

On Sat, Apr 5, 2025 at 12:40 AM Sander Devisscher ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/04_Make_country_level_predictions.R
> <https://github.com/trias-project/risk-modelling-and-mapping/pull/31#discussion_r2029516821>
> :
>
> > +#---------- Load habitat raster data -------------
> +#-------------------------------------------------
> +#ONE LAYER HAS A SLIGHTLY DIFFERENT EXTENT: CUT ALL OTHERS TO THIS EXTENT
> +habitat<-list.files((here("./data/external/habitat")),pattern='tif',full.names = T)
> +habitat_stack<-rast(habitat[c(1:5,7)]) #Distance to water (layer 6) has another extent and we're not sure whether it is correct: leave it out!
> +#habitat_stack2<-rast(habitat[6])
> +
> +#habitat_stack1<-crop(habitat_stack1, ext(habitat_stack2))
> +#habitat_stack<-c(habitat_stack1, habitat_stack2)
> +
> +
> +#-------------------------------------------------
> +#---------- Define resampling raster -------------
> +#-------------------------------------------------
> +#Used as a template when global model predictions need to be in same raster format as those for the EU model
> +resampling_raster<-rast(here("./data/external/climate/byEEA_finalRCP/belgium_rcps/rcp26/anngdd100.tif"))
>
> This file seems to be missing from the setup.
>
> Error: [rast] file does not exist: C:/Users/sander_devisscher/Documents/GitHub/risk-modelling-and-mapping/./data/external/climate/byEEA_finalRCP/belgium_rcps/rcp26/anngdd100.tif
> In addition: Warning message:
> C:/Users/sander_devisscher/Documents/GitHub/risk-modelling-and-mapping/./data/external/climate/byEEA_finalRCP/belgium_rcps/rcp26/anngdd100.tif: No such file or directory (GDAL error 4)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trias-project/risk-modelling-and-mapping/pull/31#pullrequestreview-2744225291>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC4KXBU3Q7YG2VPFGCBWRID2X4C47AVCNFSM6AAAAAB2ODEBV2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONBUGIZDKMRZGE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: <trias-project/risk-modelling-and-mapping/pull/31/review/2744225291@
> github.com>
>

damianooldoni commented:
Hi @soriadelva 👋 
I have just removed my name from the reviewers list as I am really not following this work at this moment. If you need my help, please let me know.

soriadelva commented:
> Hi @soriadelva 👋 I have just removed my name from the reviewers list as I am really not following this work at this moment. If you need my help, please let me know.

Hi @damianooldoni, no worries! I wasn't sure if you were indeed still following this work so I added you as a reviewer to notify you that the new workflow is (almost) ready, but @SanderDevisscher is thoroughly testing the scripts at the moment, so you don't need to review it :)

SanderDevisscher commented on src/02_fit_global_model.R:
```diff
@@ -0,0 +1,574 @@
+    #--------------------------------------------  
+    #Decimallon and decimalLat are converted to x and y, geometry is dropped
+    global_presabs<-global_presabs%>%
+      select(decimalLongitude, decimalLatitude)%>%
```
I get the following error:
```
Error: unable to find an inherited method for function ‘select’ for signature ‘x = "sf"’
In addition: Warning message:
In trControlCheck(x = trControl, y = target) :
  indexes not defined in trControl.  Attempting to set them ourselves, so each model in the ensemble will have the same resampling indexes.
```
I suspect this select to be the cullprit

soriadelva replied:
Can you check if it gets solved when adding `dplyr::select()`

SanderDevisscher replied:
its fixed by adding `dplyr::`

soriadelva replied:
Great, I added it to the script

soriadelva commented on src/03_fit_European_model.R:
```diff
@@ -0,0 +1,593 @@
+    eu_occ<-eu_occ%>%
+      st_coordinates()%>%
+      cbind(., eu_occ)%>%
+      select(-c(decimalLatitude, decimalLongitude))
```
One what?

damianooldoni replied:
Yes, I think @SanderDevisscher means `dplyr::`. Good practice to mention it everywhere and for every package. This will help when conflicts arise among packages and also will help everybody (future-you included) to understand the package the function comes from. Easier to find the documentation of the functions if needed.

soriadelva replied:
On it 👍 thanks @SanderDevisscher and @damianooldoni

soriadelva replied:
@SanderDevisscher dependencies have been added.

SanderDevisscher replied:
super! I'll test again with different species

SanderDevisscher reviewed (changes requested):
First the good news, with adding a just a few more `dplyr::` I was able to complete all scripts 🎉. 
So if you can go through the scripts once more and add these (and optionally as suggested by @damianooldoni all other packages except some of the base ones). 

But when I look a the results for _Lithobates catesbeianus_ & _Xenopus laevis_ I'm not 100% convinced 😅 by the results can you help me interpret them:

For both species the historical suitability matches quite good with the current situation:
Lithobates catesbeianus:
![Lithobates_catesbeianus_2427091_hist_Belgium](https://github.com/user-attachments/assets/b62195cd-ae82-4818-afd6-2961e28f5da3)

Xenopus laevis:
![Xenopus_laevis_5217334_hist_Belgium](https://github.com/user-attachments/assets/05b9cc5e-fa03-4aff-a436-c8c5a04262e8)

But all other scenarios indicate a complete loss of habitat suitability in favor for a complete different part of belgium then where the species are now 🤔, for example rcp2.6 for _Lithobates catesbeianus_:
![Lithobates_catesbeianus_2427091_rcp26_Belgium](https://github.com/user-attachments/assets/5a8aeeb5-8a61-49d7-b28d-d65bbeba03cd)
and _Xenopus laevis:_
![Xenopus_laevis_5217334_rcp26_Belgium](https://github.com/user-attachments/assets/1a513ed9-6ac5-4051-a3f7-ea3295c8b140)

Do a meet ?

SanderDevisscher reviewed (approved):
The script works 🎉 
The results for Abies alba make much more sense. This make me think the mismatch with  Lithobates & Xenopus is really due to their aquatic nature. W'll have to look a bit further into it ! But the flow works so merge it or wait for @DiederikStrubbe 's green light.

soriadelva commented:
> you should be able to match the extent of the dist2water raster to the other habitat rasters in R so you don't have to exclude it.
> @amyjsdavis, we initially did that but when we had a closer look at the dist2water layer it looked a bit off (very high values in the UK seemingly following a gradient from N to S, low values in the European mainland), so we were hesitating to include it. Would you still have the script or other documentation that we could use to recreate this layer and check the values?
